src/libraries/System.Private.CoreLib/src/System/IO/Path.cs

```diff
@@ -926,5 +926,70 @@ private static string GetRelativePath(string relativeTo, string path, StringComp
         /// Returns true if the path ends in a directory separator.
         /// </summary>
         public static bool EndsInDirectorySeparator(string path) => PathInternal.EndsInDirectorySeparator(path);
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A string containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="path"/> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentException"><paramref name="path" /> is empty.</exception>
+        public static string RemoveRedundantSegments(string path)
+        {
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+
+            return RemoveRelativeSegments(path.AsSpan());
+        }
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A read-only span containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        /// <exception cref="ArgumentException"><paramref name="path" /> is empty.</exception>
+        public static string RemoveRedundantSegments(ReadOnlySpan<char> path)
+        {
+            if (PathInternal.IsEffectivelyEmpty(path))
```

Comment:
Should this return empty output for empty input, like most other Path APIs?

E.g. `TrimEndingDirectorySeparator` is very similar to this method and it does not throw on empty paths either.

Reply:
@jkotas I decided to throw here because the `PathInternal.RemoveRelativeSegments` functions have `Debug.Asserts` checking that the passed path is not null and not empty.

If I follow your suggestion, should I remove those `Debug.Asserts` along with this exception?

Reply:
I do not think you can just remove those assert. The internal method method expects to get no-empty string on input.

Reply:
@jkotas just to make sure I understand this correctly:

There are two `RemoveRedundantSegments` functions, one that receives a `string` (A), and another that receives a `ROS<char>` (B).

For A): Instead of throwing on null or empty, make it that if `path` is `null`, return `null`; if `path` is effectively empty, return `string.Empty`. Also, decorate with the `[return: NullIfNotNull("path")]` attribute.
For B): Instead of throwing on empty, make it that if `path` is effectively empty, return `string.Empty`.

Is this correct?

Reply:
That sounds right to me - consistent with the other Path methods.

Reply:
The "effectively empty" is about resolution behavior in Windows and consistency. An empty string is equal to a string with nothing but `' '` chars as Windows will trim trailing spaces. If you ask for the directory name of  `"        "` on Windows there isn't one as windows turns it into `""`.  For `GetRelativePath` it makes no sense to find the relative location of nothing. Creating a nothing file or directory doesn't make sense, and, in fact can be problematic (confusing at the very least) as it will resolve to the parent directory we're using (when combined).

In other places we return `null` for empty/null when possible or an empty span when returning span. We should be consistent with that.

Fyi: On Framework we used to _incorrectly_ check for all Unicode space characters to make this assertion.